### PR TITLE
azureblob: add managed identity support

### PR DESCRIPTION
**Pull request by backerman**
#### What is the purpose of this change?

Adds a configuration option (`azureblob_use_msi`) to use a managed service identity when authenticating to Azure from an Azure virtual machine, which eliminates the need to manage explicit credentials. This method acquires a bearer token (valid for 24 hours) from the [Azure Instance Metadata Service](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http), and if a temporary error is returned will retry using the exponential backoff method specified in the linked documentation.

The token is not presently persisted to disk; if users are expected to invoke the `rclone` binary a large number of times in a short period that functionality would be desirable. The present implementation also does not check token validity before calling the Azure Storage blob API, so if a `rclone` invocation lasts long enough that it calls the API more than 24 hours after process start that call will fail.

#### Was the change discussed in an issue or in the forum before?

Issue #3213.

#### Checklist

- [X] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [X] I have added tests for all changes in this PR if appropriate.
- [X] I have added documentation for the changes if appropriate.
- [X] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [X] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by backerman** [comment]
Assuming the checks don't uncover any new issues that need fixing, I'll be able to check the "ready for review" box once I've added some tests to validate that the expected errors are returned when the IMDS response is 400/500-series.

File: backend/azureblob/imds.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	if isTest {
+		req.URL.Host = fmt.Sprintf("localhost:%d", testPort)
+		req.Host = req.URL.Host
+	}
+
+	// Send request
+	httpClient := fshttp.NewClient(ctx)
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		fs.Errorf(nil, "MSI is not enabled on this VM: %v", err)
```
Comment (ncw) [inline backend/azureblob/imds.go:98]: That would probably be better returned as `errors.Wrap(err, "MSI is not enabled on this VM")` and the log isn't necessary in that case

File: backend/azureblob/imds.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	default:
+		body, _ := ioutil.ReadAll(resp.Body)
+		fs.Errorf(nil, "Couldn't obtain OAuth token from IMDS; server returned status code %d and body: %v", resp.StatusCode, string(body))
+		return result, httpError{Response: resp}
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		fs.Errorf(nil, "Couldn't read IMDS response: %v", err)
+		return result, err
```
Comment (ncw) [inline backend/azureblob/imds.go:125]: This should be `errors.Wrap()` as well without the log.

File: backend/azureblob/imds.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	// Remove BOM, if any. azcopy does this so I'm following along.
+	b = bytes.TrimPrefix(b, []byte("\xef\xbb\xbf"))
+
+	// This would be a good place to persist the token if a large number of rclone
+	// invocations are being made in a short amount of time. If the token is
+	// persisted, the azureblob code will need to check for expiry before every
+	// storage API call.
+	err = json.Unmarshal(b, &result)
+	if err != nil {
+		fs.Errorf(nil, "Couldn't unmarshal IMDS response: %v", err)
```
Comment (ncw) [inline backend/azureblob/imds.go:136]: ...and here

**Review by ncw — COMMENTED** [review]
This is very nice code - thank you :-)

> Adds a configuration option (`azureblob_use_msi`) to use a managed service identity when authenticating to Azure from an Azure virtual machine, which eliminates the need to manage explicit credentials. This method acquires a bearer token (valid for 24 hours) from the [Azure Instance Metadata Service](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http), and if a temporary error is returned will retry using the exponential backoff method specified in the linked documentation.

Great

> The token is not presently persisted to disk; if users are expected to invoke the `rclone` binary a large number of times in a short period that functionality would be desirable.

Noted. If rclone users want to do this then they would be better off using the rc API.

> The present implementation also does not check token validity before calling the Azure Storage blob API, so if a `rclone` invocation lasts long enough that it calls the API more than 24 hours after process start that call will fail.

Hmm, I can see this being a potential problem. Users often kick off rclone jobs to copy TBs of data.

How difficult would it be to renew the token every few hours?

We'd then need locking around a lot more stuff though which would be rather annoying. If we want to do it then a different PR would probably be needed.

See inline for some more comments.

File: backend/azureblob/azureblob.go [hunk 4]
```diff
@@ -453,16 +488,63 @@ func NewFs(ctx context.Context, name, root string, m configmap.Mapper) (fs.Fs, e
 		}
 		pipeline := f.newPipeline(credential, azblob.PipelineOptions{Retry: azblob.RetryOptions{TryTimeout: maxTryTimeout}})
 		serviceURL = azblob.NewServiceURL(*u, pipeline)
-	case opt.Account != "" && opt.Key != "":
-		credential, err := azblob.NewSharedKeyCredential(opt.Account, opt.Key)
+	case opt.UseMSI:
+		var token adal.Token
+		trimmedClientID := strings.TrimSpace(opt.MSIClientID)
```
[the remark this answers is not in the document]
Comment (backerman) [reply backend/azureblob/azureblob.go]: Fixed.

File: backend/azureblob/imds.go (hunk 1 above)
Comment (backerman) [reply backend/azureblob/imds.go]: Changed these ones as suggested.

File: backend/azureblob/imds.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	return fmt.Sprintf("HTTP error %v (%v)", e.Response.StatusCode, e.Response.Status)
+}
+
+// GetMSIToken attempts to obtain an MSI token from the Azure Instance
+// Metadata Service.
+func GetMSIToken(ctx context.Context, identity *userMSI) (adal.Token, error) {
+	// Attempt to get an MSI token; silently continue if unsuccessful.
+	// Future work: Allow user to assert that an MSI token must be
+	// available.
+	// This code has been lovingly stolen from azcopy's OAuthTokenManager.
```
[the remark this answers is not in the document]
Comment (backerman) [reply backend/azureblob/imds.go:55]: I'll look around again, but I don't think so - the sample Go for getting the token does it with `net/http` as well.

**Comment by backerman** [comment]
> How difficult would it be to renew the token every few hours?

Looks like it's just including a [`TokenRefresher`](https://godoc.org/github.com/Azure/azure-storage-blob-go/azblob#TokenRefresher) function where I've currently got a `nil`, so I'll do that. Shouldn't be too difficult and the pipeline package is already dealing with the concurrency issues there.

File: backend/azureblob/azureblob.go [hunk 6]
```diff
@@ -453,16 +488,60 @@ func NewFs(ctx context.Context, name, root string, m configmap.Mapper) (fs.Fs, e
...
-			return nil, errors.Wrapf(err, "Failed to parse credentials")
+			return nil, errors.Wrapf(err, "Failed to acquire MSI token")
 		}
 
 		u, err = url.Parse(fmt.Sprintf("https://%s.%s", opt.Account, opt.Endpoint))
 		if err != nil {
 			return nil, errors.Wrap(err, "failed to make azure storage url from account and endpoint")
 		}
+
+		credential := azblob.NewTokenCredential(token.AccessToken, nil)
```
Comment (backerman) [inline backend/azureblob/azureblob.go:537]: This is where the `TokenRefresher` goes.

**Comment by backerman** [comment]
Okay, added token refresh support, but I'm not sure how to fix this conflict warning; there doesn't even appear to be a conflicta.

File: backend/azureblob/imds.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (backerman) [reply backend/azureblob/imds.go:55]: Looks like there is in fact `auth.NewAuthorizerFromEnvironment` which is the droid we're looking for.

[the remark this answers is not in the document]
Comment (backerman) [reply backend/azureblob/imds.go:55]: And that's actually only for the main Azure Go SDK; the blob storage SDK doesn't use a compatible interface.

**Comment by ncw** [comment]
Sorry for the delay responding. 

This looks ready to merge now - would you agree?

Thanks

Nick

**Comment by backerman** [comment]
> This looks ready to merge now - would you agree?

I just rebased to master and modified the message generated when no authentication type is available (getting a bit long to list them individually). Looks good to me once the checks pass.

**Comment by ncw** [comment]
That looks great now - thank you for the great feature :-)

I'll merge this now.

**Comment by ncw** [comment]
Hi @backerman 

I just fixed a small oopsie in this code that the integration tests threw up cb97c2b0d39b38e1ae4f424a62503717495fa36c - that isn't a problem - these things happen!

What I was wondering about while looking at that was this in the new code you added. `opt.Account` and `opt.Endpoint` probably won't be set here will they? So maybe we don't need to make `u` at all.

https://github.com/rclone/rclone/blob/cb97c2b0d39b38e1ae4f424a62503717495fa36c/backend/azureblob/azureblob.go#L617-L620

We use it here - maybe we can pass in empty string or nil?

https://github.com/rclone/rclone/blob/cb97c2b0d39b38e1ae4f424a62503717495fa36c/backend/azureblob/azureblob.go#L645

What do you think?

**Comment by backerman** [comment]
> Hi @backerman
> 
> I just fixed a small oopsie in this code that the integration tests threw up [cb97c2b](https://github.com/rclone/rclone/commit/cb97c2b0d39b38e1ae4f424a62503717495fa36c) - that isn't a problem - these things happen!
> 
> What I was wondering about while looking at that was this in the new code you added. `opt.Account` and `opt.Endpoint` probably won't be set here will they? So maybe we don't need to make `u` at all.

They should be set (at least `Account`) in the configuration for any authentication method; it would probably be a good idea to extract them from the switch statement, though.

**Comment by ncw** [comment]
> > I just fixed a small oopsie in this code that the integration tests threw up [cb97c2b](https://github.com/rclone/rclone/commit/cb97c2b0d39b38e1ae4f424a62503717495fa36c) - that isn't a problem - these things happen!
> > What I was wondering about while looking at that was this in the new code you added. `opt.Account` and `opt.Endpoint` probably won't be set here will they? So maybe we don't need to make `u` at all.
> 
> They should be set (at least `Account`) in the configuration for any authentication method; it would probably be a good idea to extract them from the switch statement, though.

Ah OK!

I'm happy to leave it how it is then. Factoring it would be nice but not all the if branches need it so it probably isn't worth it.

**Pull request merged**